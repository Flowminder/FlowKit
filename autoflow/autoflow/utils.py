# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.

"""
Utility functions.
"""

import collections
import json
import nbconvert
import nbformat
import networkx as nx
import pendulum
from contextlib import contextmanager
from get_secret_or_env_var import getenv
from hashlib import md5
from pathlib import Path
from sh import asciidoctor_pdf
from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker
from tempfile import TemporaryDirectory
from typing import Any, Dict, Iterator, Optional, OrderedDict, Set, Tuple, Union


def get_output_filename(input_filename: str, tag: str = "") -> str:
    """
    Given an input filename, construct an output filename with the same extension,
    with a timestamp (and optionally a tag) added to the stem.

    Parameters
    ----------
    input_filename : str
        Input filename
    tag : str, optional
        A tag to append to the file stem
    
    Returns
    -------
    str
        Output filename
    """
    input_filename = Path(input_filename)
    now_string = pendulum.now("utc").format("YYYYMMDDTHHmmss[Z]")
    tag = f"__{tag}" if tag != "" else tag
    return f"{input_filename.stem}{tag}__{now_string}{input_filename.suffix}"


def get_params_hash(parameters: Dict[str, Any]) -> str:
    """
    Generate a md5 hash string from a dictionary of parameters.
    The dictionary is dumped to a json string before hashing.

    Parameters
    ----------
    parameters : dict
        Dictionary of parameters to use for creating the hash.
    
    Returns
    -------
    str
        md5 hash of the parameters dict
    """
    return md5(
        json.dumps(dict(parameters), sort_keys=True, default=str).encode()
    ).hexdigest()


def get_session(db_uri: str) -> "sqlalchemy.orm.session.Session":
    """
    Create a sqlalchemy session.

    Parameters
    ----------
    db_uri : str
        Database URI
    
    Returns
    -------
    Session
        A sqlalchemy session
    """
    # TODO: This seems like the wrong place to be reading a secret / env var,
    # but we can't put a docker secret in the prefect config.
    full_db_uri = db_uri.format(getenv("AUTOFLOW_DB_PASSWORD", ""))
    engine = create_engine(full_db_uri)
    return sessionmaker(bind=engine)()


@contextmanager
def session_scope(db_uri: str) -> Iterator["sqlalchemy.orm.session.Session"]:
    """
    Provide a scoped sqlalchemy session for interacting with the database.

    Parameters
    ----------
    db_uri : str
        Database URI
    
    Yields
    ------
    Session
        A sqlalchemy session
    """
    session = get_session(db_uri)
    try:
        yield session
        session.commit()
    except:
        session.rollback()
        raise
    finally:
        session.close()


def make_json_serialisable(obj: Any) -> Union[dict, list, str, int, float, bool, None]:
    """
    Helper function to convert an object's type so that it can be serialised using
    the default json serialiser. Non-serialisable types are converted to strings.

    Parameters
    ----------
    obj : any
        Object to be converted.
    
    Returns
    -------
    dict, list, str, int, float, bool or None
        The same object after dumping to json (converting to str if necessary) and loading
    """
    return json.loads(json.dumps(obj, default=str))


def get_additional_parameter_names_for_notebooks(
    notebooks: Dict[str, Dict[str, Any]],
    reserved_parameter_names: Optional[Set[str]] = None,
) -> Set[str]:
    """
    Extract parameter names from a list of notebook task specifications.

    Parameters
    ----------
    notebooks : dict
        Dictionary of dictionaries describing notebook tasks.
    reserved_parameter_names : set of str, optional
        Names of parameters used within workflow.
    
    Returns
    -------
    set of str
        Names of parameters used by notebooks which must be passed to a workflow.
    """
    # Parameters requested in notebooks
    notebook_parameter_names = set.union(
        *[set(notebook["parameters"].values()) for notebook in notebooks.values()]
    )
    # Parameters available to notebooks, which are not passed externally as flow parameters
    internal_parameter_names = (reserved_parameter_names or set()).union(
        notebooks.keys()
    )
    # Additional parameters required for notebooks
    additional_parameter_names_for_notebooks = notebook_parameter_names.difference(
        internal_parameter_names
    )
    return additional_parameter_names_for_notebooks


def sort_notebooks(
    notebooks: Dict[str, Dict[str, Any]]
) -> OrderedDict[str, Dict[str, Any]]:
    """
    Perform a topological sort on a dictionary of notebook task specifications.

    Parameters
    ----------
    notebooks : dict
        Dictionary of dictionaries describing notebook tasks.
    
    Returns
    -------
    OrderedDict
        Ordered dict of notebook task dicts, ordered so that no notebook depends on another
        notebook that comes after it.
    
    Raises
    ------
    ValueError
        If the notebook specifications contain circular dependencies.
    """
    notebooks_graph = nx.DiGraph(
        {
            key: [
                value
                for value in notebooks[key]["parameters"].values()
                if value in notebooks
            ]
            for key in notebooks
        }
    ).reverse()
    try:
        sorted_notebook_keys = list(nx.topological_sort(notebooks_graph))
    except nx.NetworkXUnfeasible:
        raise ValueError("Notebook specifications contain circular dependencies.")

    return collections.OrderedDict(
        (key, notebooks[key]) for key in sorted_notebook_keys
    )


def notebook_to_asciidoc(
    notebook_path: str, asciidoc_template_path: Optional[str] = None
) -> Tuple[str, "nbconvert.exporters.exporter.ResourcesDict"]:
    """
    Convert a Jupyter notebook to ASCIIDoc, using nbconvert.

    Parameters
    ----------
    notebook_path : str
        Path to notebook
    asciidoc_template_path : str, optional
        Path to a template file to use for the conversion.
        If not provided, the default nbconvert asciidoc template will be used.

    Returns
    -------
    str
        The resulting converted notebook.
    ResourcesDict
        Dictionary of resources used prior to and during the conversion process.
    """
    with open(notebook_path) as nb_file:
        nb_read = nbformat.read(nb_file, as_version=4)

    if asciidoc_template_path is None:
        exporter = nbconvert.ASCIIDocExporter()
    else:
        exporter = nbconvert.ASCIIDocExporter(template_file=asciidoc_template_path)
    body, resources = exporter.from_notebook_node(nb_read)

    return body, resources


def asciidoc_to_pdf(
    body: str, resources: "nbconvert.exporters.exporter.ResourcesDict", output_path: str
) -> None:
    """
    Convert an ASCIIDoc document to PDF, using asciidoctor-pdf.

    Parameters
    ----------
    body : str
        Content of the ASCIIDoc document
    resources : ResourcesDict
        Dictionary of resources, as returned from nbconvert exporter
    """
    with TemporaryDirectory() as tmpdir:
        with open(f"{tmpdir}/tmp.asciidoc", "w") as f_tmp:
            f_tmp.write(body)
        for fname, content in resources["outputs"].items():
            with open(f"{tmpdir}/{fname}", "wb") as fout:
                fout.write(content)
        asciidoctor_pdf(f"{tmpdir}/tmp.asciidoc", o=output_path)
