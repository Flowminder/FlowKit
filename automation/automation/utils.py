# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.

import json
import datetime
import pendulum
from hashlib import md5
from pathlib import Path
from typing import Union, Dict, Any, List, Sequence, Set, Tuple, Optional
from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker


def get_output_filename(input_filename: str, tag: str = "") -> str:
    """
    Given an input filename, construct an output filename with the same extension,
    with a timestamp (and optionally a label) added to the stem.

    Parameters
    ----------
    input_filename : str
        Input filename
    tag : str, optional
        A tag to append to the file stem
    
    Returns
    -------
    str
        Output filename
    """
    input_filename = Path(input_filename)
    now_string = pendulum.now("utc").format("YYYYMMDDTHHmmss[Z]")
    return f"{input_filename.stem}{tag}__{now_string}{input_filename.suffix}"


def get_params_hash(parameters: Dict[str, Any]) -> str:
    """
    Generate a md5 hash string from a dictionary of parameters.
    The dictionary os dumped to a json string before hashing.

    Parameters
    ----------
    parameters : dict
        Dictionary of parameters to use for creating the hash.
    
    Returns
    -------
    str
        md5 hash of the parameters dict
    """
    return md5(
        json.dumps(dict(parameters), sort_keys=True, default=str).encode()
    ).hexdigest()


date_or_offset = Union[int, datetime.date]  # Type alias for elements of date stencils
stencil_type_alias = List[
    Union[date_or_offset, List[date_or_offset]]
]  # Type alias for date stencils


class InvalidDatePairError(ValueError):
    """
    Custom error to raise if populating a date stencil results in a date pair with start_date > end_date.
    """

    pass


def offset_to_date(
    offset: date_or_offset, reference_date: datetime.date
) -> pendulum.Date:
    """
    Return a date corresponding to the offset from a reference date.

    Parameters
    ----------
    offset : int or datetime.date
        Either an integer number of days offset from reference date, or a date object.
        If a date object, this date will be returned.
    reference_date : datetime.date
        Date to calculate the offset relative to.

    Returns
    -------
    pendulum.Date
        reference_date + offset (if offset is an integer), or offset (if offset is a date).
    
    Raises
    ------
    TypeError
        If type(offset) is not either int or datetime.date
    """
    if isinstance(offset, datetime.date):
        date_from_offset = pendulum.date(offset.year, offset.month, offset.day)
    elif isinstance(offset, int):
        date_from_offset = pendulum.date(
            reference_date.year, reference_date.month, reference_date.day
        ).add(days=offset)
    else:
        raise TypeError(
            f"Invalid type for offset: expected 'date' or 'int', not '{type(offset)}'"
        )
    return date_from_offset


def stencil_to_date_pairs(
    stencil: stencil_type_alias, reference_date: datetime.date
) -> List[Tuple[pendulum.Date, pendulum.Date]]:
    """
    Given a stencil of dates, date offsets and/or date intervals, and a reference
    date to calculate the offsets relative to, return a list of date pairs representing
    date intervals (inclusive of both limits).

    Parameters
    ----------
    stencil : list of datetime.date, int and/or pairs of date/int
        List of elements defining dates or date intervals.
        Each element can be:
            - a date object corresponding to an absolute date,
            - an int corresponding to an offset (in days) relative to reference_date,
            - a length-2 list [start, end] of dates or offsets, corresponding to a
              date interval (inclusive of both limits).
    reference_date : datetime.date
        Date to calculate offsets relative to.
    
    Returns
    -------
    list of tuple (pendulum.Date, pendulum.Date)
        List of pairs of date objects, each representing a date interval.
    
    Raises
    ------
    TypeError
        If elements of stencil have the wrong type
    ValueError
        If list elements of stencil do not have length 2
    InvalidDatePairError
        If the stencil results in a date pair with start_date > end_date
    """
    date_pairs = []
    for element in stencil:
        if isinstance(element, list):
            if len(element) != 2:
                raise ValueError(
                    "Expected date interval to be a list of length 2 (in format [start, end]), "
                    "but got list of length {len(element)}."
                )
            start_date = offset_to_date(element[0], reference_date)
            end_date = offset_to_date(element[1], reference_date)
            if start_date > end_date:
                raise InvalidDatePairError(
                    f"Stencil contains invalid date pair ({start_date}, {end_date}) for reference date {reference_date}."
                )
            date_pairs.append((start_date, end_date))
        else:
            date = offset_to_date(element, reference_date)
            date_pairs.append((date, date))
    return date_pairs


def stencil_to_set_of_dates(
    stencil: stencil_type_alias, reference_date: datetime.date
) -> Set[pendulum.Date]:
    """
    Given a stencil of dates, date offsets and/or date intervals, and a reference
    date to calculate the offsets relative to, return the corresponding set of dates.

    Parameters
    ----------
    stencil : list of datetime.date, int and/or pairs of date/int
        List of elements defining dates or date intervals.
        Each element can be:
            - a date object corresponding to an absolute date,
            - an int corresponding to an offset (in days) relative to reference_date,
            - a length-2 list [start, end] of dates or offsets, corresponding to a
              date interval (inclusive of both limits).
    reference_date : datetime.date
        Date to calculate offsets relative to.

    Returns
    -------
    set of pendulum.Date
        Set of dates represented by the stencil
    """
    date_pairs = stencil_to_date_pairs(stencil, reference_date)
    dates = set().union(*[pendulum.period(pair[0], pair[1]) for pair in date_pairs])
    return dates


def dates_are_available(
    stencil: stencil_type_alias,
    reference_date: datetime.date,
    available_dates: Sequence[datetime.date],
) -> bool:
    """
    Check whether all dates represented by a stencil for a particular date are available.

    Parameters
    ----------
    stencil : list of datetime.date, int and/or pairs of date/int
        List of elements defining dates or date intervals.
        Each element can be:
            - a date object corresponding to an absolute date,
            - an int corresponding to an offset (in days) relative to reference_date,
            - a length-2 list [start, end] of dates or offsets, corresponding to a
              date interval (inclusive of both limits).
    reference_date : datetime.date
        Date to calculate offsets relative to.
    available_dates : list of datetime.date
        List of available dates
    
    Returns
    -------
    bool
        True if all dates are available, False otherwise.
    
    Notes
    -----

    If the stencil is not valid for the given reference date (i.e. contains invalid
    date pairs), this function will return False.
    """
    try:
        dates_from_stencil = stencil_to_set_of_dates(
            stencil, reference_date=reference_date
        )
    except InvalidDatePairError:
        return False
    return dates_from_stencil.issubset(set(available_dates))


def get_session(db_uri: str) -> "sqlalchemy.orm.session.Session":
    """
    Create a sqlalchemy session.

    Parameters
    ----------
    db_uri : str
        Database URI

    Returns
    -------
    Session
        A sqlalchemy session
    """
    engine = create_engine(db_uri)
    return sessionmaker(bind=engine)()


def make_json_serialisable(obj: Any) -> Union[dict, list, str, int, float, bool, None]:
    """
    Helper function to convert an object's type so that it can be serialised using
    the default json serialiser. non-serialisable types are converted to strings.

    Parameters
    ----------
    obj : any
        Object to be converted.
    
    Returns
    -------
    dict, list, str, int, float, bool or None
        The same object after dumping to json (converting to str if necessary) and loading
    """
    return json.loads(json.dumps(obj, default=str))


def get_parameter_names(
    notebooks: List[Dict[str, Any]], reserved_parameter_names: Optional[Set[str]] = None
) -> Tuple[Set[str], Set[str]]:
    """
    Extract parameter names from a list of notebook task specifications.

    Parameters
    ----------
    notebooks : list of dict
        List of dictionaries describing notebook tasks.
    reserved_parameter_names : set of str, optional
        Names of parameters used within workflow, which cannot be used as notebook labels.
    
    Returns
    -------
    notebook_labels : set of str
        Notebook labels, which can be used as parameter names in other notebook tasks
    additional_parameter_names_for_notebooks : set of str
        Names of parameters used by notebooks which are not either reserved parameter names or notebook labels
    """
    if reserved_parameter_names is None:
        reserved_parameter_names = set()
    # Labels for notebook tasks
    notebook_labels = set(notebook["label"] for notebook in notebooks)
    forbidden_labels = notebook_labels.intersection(reserved_parameter_names)
    if forbidden_labels:
        raise ValueError(
            f"Notebook labels {forbidden_labels} are forbidden for this workflow. "
            f"Reserved parameter names are {reserved_parameter_names}."
        )
    # Parameters requested in notebooks
    notebook_parameter_names = set.union(
        *[set(notebook["parameters"].values()) for notebook in notebooks]
    )
    # Additional parameters required for notebooks
    additional_parameter_names_for_notebooks = notebook_parameter_names.difference(
        reserved_parameter_names
    ).difference(notebook_labels)
    return notebook_labels, additional_parameter_names_for_notebooks
