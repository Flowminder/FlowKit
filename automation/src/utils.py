from pathlib import Path

def get_output_filename(input_path, output_dir, label):
    input_path = Path(input_path)
    return f"{output_dir}/{input_path.stem}_{label}{input_path.suffix}"