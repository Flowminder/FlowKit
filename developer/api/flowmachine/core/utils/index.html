



<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      <meta http-equiv="x-ua-compatible" content="ie=edge">
      
      
      
      
        <meta name="lang:clipboard.copy" content="Copy to clipboard">
      
        <meta name="lang:clipboard.copied" content="Copied to clipboard">
      
        <meta name="lang:search.language" content="en">
      
        <meta name="lang:search.pipeline.stopwords" content="True">
      
        <meta name="lang:search.pipeline.trimmer" content="True">
      
        <meta name="lang:search.result.none" content="No matching documents">
      
        <meta name="lang:search.result.one" content="1 matching document">
      
        <meta name="lang:search.result.other" content="# matching documents">
      
        <meta name="lang:search.tokenizer" content="[\s\-]+">
      
      <link rel="shortcut icon" href="../../../../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.0.4, mkdocs-material-3.2.0">
    
    
      
        <title>utils</title>
      
    
    
      <link rel="stylesheet" href="../../../../../assets/stylesheets/application.572ca0f0.css">
      
        <link rel="stylesheet" href="../../../../../assets/stylesheets/application-palette.22915126.css">
      
      
        
        
        <meta name="theme-color" content="">
      
    
    
      <script src="../../../../../assets/javascripts/modernizr.8c900955.js"></script>
    
    
      
        <link href="https://fonts.gstatic.com" rel="preconnect" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700|Roboto+Mono">
        <style>body,input{font-family:"Roboto","Helvetica Neue",Helvetica,Arial,sans-serif}code,kbd,pre{font-family:"Roboto Mono","Courier New",Courier,monospace}</style>
      
    
    <link rel="stylesheet" href="../../../../../assets/fonts/material-icons.css">
    
    
      <link rel="stylesheet" href="../../../../../stylesheets/extra.css">
    
      <link rel="stylesheet" href="https://unpkg.com/mermaid@7.1.2/dist/mermaid.css">
    
    
  </head>
  
    
    
    <body dir="ltr" data-md-color-primary="#095798" data-md-color-accent="#2977b8">
  
    <svg class="md-svg">
      <defs>
        
        
          <svg xmlns="http://www.w3.org/2000/svg" width="416" height="448"
    viewBox="0 0 416 448" id="__github">
  <path fill="currentColor" d="M160 304q0 10-3.125 20.5t-10.75 19-18.125
        8.5-18.125-8.5-10.75-19-3.125-20.5 3.125-20.5 10.75-19 18.125-8.5
        18.125 8.5 10.75 19 3.125 20.5zM320 304q0 10-3.125 20.5t-10.75
        19-18.125 8.5-18.125-8.5-10.75-19-3.125-20.5 3.125-20.5 10.75-19
        18.125-8.5 18.125 8.5 10.75 19 3.125 20.5zM360
        304q0-30-17.25-51t-46.75-21q-10.25 0-48.75 5.25-17.75 2.75-39.25
        2.75t-39.25-2.75q-38-5.25-48.75-5.25-29.5 0-46.75 21t-17.25 51q0 22 8
        38.375t20.25 25.75 30.5 15 35 7.375 37.25 1.75h42q20.5 0
        37.25-1.75t35-7.375 30.5-15 20.25-25.75 8-38.375zM416 260q0 51.75-15.25
        82.75-9.5 19.25-26.375 33.25t-35.25 21.5-42.5 11.875-42.875 5.5-41.75
        1.125q-19.5 0-35.5-0.75t-36.875-3.125-38.125-7.5-34.25-12.875-30.25-20.25-21.5-28.75q-15.5-30.75-15.5-82.75
        0-59.25 34-99-6.75-20.5-6.75-42.5 0-29 12.75-54.5 27 0 47.5 9.875t47.25
        30.875q36.75-8.75 77.25-8.75 37 0 70 8 26.25-20.5
        46.75-30.25t47.25-9.75q12.75 25.5 12.75 54.5 0 21.75-6.75 42 34 40 34
        99.5z" />
</svg>
        
      </defs>
    </svg>
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" data-md-component="overlay" for="__drawer"></label>
    
      <a href="#flowmachinecoreutils" tabindex="1" class="md-skip">
        Skip to content
      </a>
    
    
      <header class="md-header" data-md-component="header">
  <nav class="md-header-nav md-grid">
    <div class="md-flex">
      <div class="md-flex__cell md-flex__cell--shrink">
        <a href="../../../../.." title="FlowKit" class="md-header-nav__button md-logo">
          
            <i class="md-icon"></i>
          
        </a>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        <label class="md-icon md-icon--menu md-header-nav__button" for="__drawer"></label>
      </div>
      <div class="md-flex__cell md-flex__cell--stretch">
        <div class="md-flex__ellipsis md-header-nav__title" data-md-component="title">
          
            
              <span class="md-header-nav__topic">
                FlowKit
              </span>
              <span class="md-header-nav__topic">
                utils
              </span>
            
          
        </div>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        
          
        
      </div>
      
        <div class="md-flex__cell md-flex__cell--shrink">
          <div class="md-header-nav__source">
            


  


  <a href="https://github.com/Flowminder/FlowKit" title="Go to repository" class="md-source" data-md-source="github">
    
      <div class="md-source__icon">
        <svg viewBox="0 0 24 24" width="24" height="24">
          <use xlink:href="#__github" width="24" height="24"></use>
        </svg>
      </div>
    
    <div class="md-source__repository">
      Flowminder/FlowKit
    </div>
  </a>

          </div>
        </div>
      
    </div>
  </nav>
</header>
    
    <div class="md-container">
      
        
      
      
        

<nav class="md-tabs" data-md-component="tabs">
  <div class="md-tabs__inner md-grid">
    <ul class="md-tabs__list">
      
        
      
        
      
        
      
        
      
        
      
        
      
    </ul>
  </div>
</nav>
      
      <main class="md-main">
        <div class="md-main__inner md-grid" data-md-component="container">
          
            
              <div class="md-sidebar md-sidebar--primary" data-md-component="navigation">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    <nav class="md-nav md-nav--primary" data-md-level="0">
  <label class="md-nav__title md-nav__title--site" for="__drawer">
    <a href="../../../../.." title="FlowKit" class="md-nav__button md-logo">
      
        <i class="md-icon"></i>
      
    </a>
    FlowKit
  </label>
  
    <div class="md-nav__source">
      


  


  <a href="https://github.com/Flowminder/FlowKit" title="Go to repository" class="md-source" data-md-source="github">
    
      <div class="md-source__icon">
        <svg viewBox="0 0 24 24" width="24" height="24">
          <use xlink:href="#__github" width="24" height="24"></use>
        </svg>
      </div>
    
    <div class="md-source__repository">
      Flowminder/FlowKit
    </div>
  </a>

    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      


  <li class="md-nav__item">
    <a href="../../../../../introduction/" title="Introduction" class="md-nav__link">
      Introduction
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../../../../../install/" title="Installation" class="md-nav__link">
      Installation
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../../../../../flowclient/" title="Analyst" class="md-nav__link">
      Analyst
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../../../../roadmap/" title="Developer" class="md-nav__link">
      Developer
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../../../../../administrator/cache_management/" title="Administrator" class="md-nav__link">
      Administrator
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../../../../../license/" title="License" class="md-nav__link">
      License
    </a>
  </li>

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              <div class="md-sidebar md-sidebar--secondary" data-md-component="toc">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary">
  
  
    
  
  
    <label class="md-nav__title" for="__toc">Table of contents</label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#pretty_sql" title="pretty_sql" class="md-nav__link">
    pretty_sql
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#parameters" title="Parameters" class="md-nav__link">
    Parameters
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns" title="Returns" class="md-nav__link">
    Returns
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
      
      
      
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content">
            <article class="md-content__inner md-typeset">
              
                
                
                <h1 id="flowmachinecoreutils">flowmachine.core.utils<a class="headerlink" href="#flowmachinecoreutils" title="Permanent link">&para;</a></h1>
<p>Source: <a href="https://github.com/Flowminder/FlowKit/tree/docs-update/flowmachine/flowmachine/core/utils.py#L0">flowmachine/core/utils.py</a></p>
<p>Utility functions that are used only by the core code and are not therefore user facing.</p>
<h2 id="pretty_sql">pretty_sql<a class="headerlink" href="#pretty_sql" title="Permanent link">&para;</a></h2>
<p><div class="codehilite"><pre><span></span><span class="n">pretty_sql</span><span class="p">(</span>
    <span class="n">sql</span><span class="p">,</span>
    <span class="n">compact_lists_margin</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
    <span class="n">split_string_literals_threshold</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
    <span class="n">special_functions</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
    <span class="n">comma_at_eoln</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
Source: <a href="https://github.com/Flowminder/FlowKit/tree/docs-update/flowmachine/flowmachine/core/utils.py#L18">flowmachine/core/utils.py</a></p>
<p>Prettify and validate the syntax of an SQL query, using pglast</p>
<h3 id="parameters">Parameters<a class="headerlink" href="#parameters" title="Permanent link">&para;</a></h3>
<ul>
<li>
<p><code>sql</code>: <code>str</code></p>
<p>SQL to prettyify and validate</p>
</li>
<li>
<p><code>compact_lists_margin</code>: <code>int</code>, default <code>0</code></p>
<p>Use compact form for lists shorter than this</p>
</li>
<li>
<p><code>split_string_literals_threshold</code>: <code>int</code>, default <code>0</code></p>
<p>Split strings (in the sql) longer than this threshold</p>
</li>
<li>
<p><code>special_functions</code>: <code>bool</code>, default <code>True</code></p>
<p>Translate some special functions to their more commonly used forms</p>
</li>
<li>
<p><code>comma_at_eoln</code></p>
</li>
</ul>
<h3 id="returns">Returns<a class="headerlink" href="#returns" title="Permanent link">&para;</a></h3>
<ul>
<li>
<p><code>str</code></p>
<p>The prettified string.</p>
</li>
</ul>
                
                  
                
              
              
                


              
            </article>
          </div>
        </div>
      </main>
      
        
<footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
        powered by
        <a href="https://www.mkdocs.org">MkDocs</a>
        and
        <a href="https://squidfunk.github.io/mkdocs-material/">
          Material for MkDocs</a>
      </div>
      
        
  <div class="md-footer-social">
    <link rel="stylesheet" href="../../../../../assets/fonts/font-awesome.css">
    
      <a href="https://github.com/Flowminder" class="md-footer-social__link fa fa-github"></a>
    
      <a href="https://twitter.com/Flowminder" class="md-footer-social__link fa fa-twitter"></a>
    
  </div>

      
    </div>
  </div>
</footer>
      
    </div>
    
      <script src="../../../../../assets/javascripts/application.b41f3d20.js"></script>
      
      <script>app.initialize({version:"1.0.4",url:{base:"../../../../.."}})</script>
      
        <script src="https://unpkg.com/mermaid@7.1.2/dist/mermaid.min.js"></script>
      
    
    
      
    
  </body>
</html>