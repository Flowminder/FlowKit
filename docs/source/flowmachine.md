# FlowMachine

FlowMachine is a Python toolkit for the analysis of CDR data. It is essentially a python wrapper for postgres SQL queries, including geographical queries with postgis. All of the classes in flowmachine define an SQL query under the hood.

## Documentation

Documentation for FlowMachine can be found [here](../developer/api/flowmachine).
