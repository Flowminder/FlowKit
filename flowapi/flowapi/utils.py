# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.


from quart import request
from quart.exceptions import HTTPException


async def get_query_parameters_from_flowmachine(*, query_id) -> dict:
    """
    Get the parameters of a query from flowmachine.

    Parameters
    ----------
    query_id : str
        ID of the query to get params for

    Returns
    -------
    dict
        Dictionary containing the query's original parameters

    Raises
    ------
    HTTPException
        404 if the query id is not known.

    """
    request.socket.send_json(
        {
            "request_id": request.request_id,
            "action": "get_query_params",
            "params": {"query_id": query_id},
        }
    )
    reply = await request.socket.recv_json()
    if reply["status"] == "error":
        raise HTTPException(
            description=f"Unknown query ID '{query_id}'",
            name="Query ID not found",
            status_code=404,
        )
    return reply["payload"]["query_params"]
