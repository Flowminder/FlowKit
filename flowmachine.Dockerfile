# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.

FROM python:3.8-slim

ARG SOURCE_VERSION=0+unknown
ENV SOURCE_VERSION=${SOURCE_VERSION}
ENV SOURCE_TREE=FlowKit-${SOURCE_VERSION}
WORKDIR /${SOURCE_TREE}/flowmachine
COPY ./flowmachine/Pipfile* ./
RUN apt-get update && \
        apt-get install -y --no-install-recommends git && \
        pip install --no-cache-dir pipenv && pipenv install --clear --deploy && \
        apt-get -y remove git && \
        apt purge -y --auto-remove && \
        rm -rf /var/lib/apt/lists/*
COPY . /${SOURCE_TREE}/
RUN apt-get update && \
        apt-get install -y --no-install-recommends git && \
        pipenv run python setup.py install && \
        apt-get -y remove git && \
        apt purge -y --auto-remove && \
        rm -rf /var/lib/apt/lists/*
CMD ["pipenv", "run", "flowmachine"]
# FlowDB has a default role named flowmachine for use with the flowmachine server
# when starting the container with a different user, that user must be in the flowmachine
# role
ENV FLOWMACHINE_FLOWDB_USER=flowmachine
