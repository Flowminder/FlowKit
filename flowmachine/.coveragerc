[run]
branch = True

[report]
fail_under = 90

omit =
    */python?.?/*
    */venv/*
    */site-packages/nose/*
    */tests/*
    */_version.py
    */versioneer.py
    *__init__*
    flowmachine/core/server/server.py

exclude_lines =
    if __name__ == '__main__':

show_missing = True
