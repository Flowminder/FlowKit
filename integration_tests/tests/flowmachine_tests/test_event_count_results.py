# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.

from flowmachine.utils import pretty_sql

from approvaltests.approvals import verify
from flowmachine.core import CustomQuery
from flowmachine.features import EventCount


def test_event_count_1_sql(diff_reporter):
    """
    Event count query with non-default parameters returns the expected data.
    """
    ec = EventCount(start="2016-01-02", stop="2016-01-04", tables=None)
    sql = pretty_sql(ec.get_query())
    verify(sql, diff_reporter)


def test_event_count_1_df(get_dataframe, diff_reporter):
    """
    Event count query with non-default parameters returns the expected data.
    """
    ec = EventCount(start="2016-01-02", stop="2016-01-04", tables=None)
    df = get_dataframe(ec)
    verify(df.to_csv(), diff_reporter)
