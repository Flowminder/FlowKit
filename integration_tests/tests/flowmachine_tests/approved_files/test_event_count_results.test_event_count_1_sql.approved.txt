SELECT subscriber,
       count(*) AS value
FROM (SELECT events.calls.msisdn AS subscriber
      FROM events.calls
      WHERE (events.calls.datetime >= '2016-01-02 00:00:00')
        AND (events.calls.datetime < '2016-01-04 00:00:00')

      UNION ALL

      SELECT events.sms.msisdn AS subscriber
      FROM events.sms
      WHERE (events.sms.datetime >= '2016-01-02 00:00:00')
        AND (events.sms.datetime < '2016-01-04 00:00:00')) AS u
GROUP BY subscriber